toolchain/base/index_base.h

```diff
@@ -52,78 +53,32 @@ struct IndexBase : public IdBase {
   using IdBase::IdBase;
 };
 
-// Equality comparison for both IdBase and IndexBase.
-template <
-    typename IndexType,
-    typename std::enable_if_t<std::is_base_of_v<IdBase, IndexType>>* = nullptr>
+// Equality comparison for both IdBase and IndexBase, including the index type
+// on either the LHS or RHS with something convertible to it. Inequality
+// automatically uses these in C++20.
+template <typename IndexType>
+  requires std::derived_from<IndexType, IdBase>
 auto operator==(IndexType lhs, IndexType rhs) -> bool {
   return lhs.index == rhs.index;
 }
-template <
-    typename IndexType,
-    typename std::enable_if_t<std::is_base_of_v<IdBase, IndexType>>* = nullptr>
-auto operator!=(IndexType lhs, IndexType rhs) -> bool {
-  return lhs.index != rhs.index;
-}
-
-template <
-    typename IndexType, typename RHSType,
-    typename std::enable_if_t<std::is_base_of_v<IdBase, IndexType>>* = nullptr,
-    typename std::enable_if_t<std::is_convertible_v<RHSType, IndexType>>* =
-        nullptr>
+template <typename IndexType, typename RHSType>
+  requires std::derived_from<IndexType, IdBase> &&
+           std::convertible_to<RHSType, IndexType>
```

Comment:
Not arguing for a change here, just a data point for the style rule: I think type-constraints in the template parameter list would be significantly easier to read than this requires-clause. `template <std::derived_from<IdBase> IndexType, std::convertible_to<IndexType> LHSType>` is clearly looking for one type that is derived from `IdBase` and another type that is convertible to that first type, whereas this formulation makes it pretty unclear which type we expect to be derived from which and which type we expect to be convertible to which.

The standard library concepts are named in such a way that use in a template parameter list is readable, which I think will tend to make use in a requires-clause confusing. This isn't a new problem -- `is_base_of` and `is_convertible` have the same lack of clarity, and there are some especially bad examples like `is_convertible` and `is_constructible` making opposite choices (they model "is convertible *to*" versus "is constructible *from*") -- but use of a type constraint would solve this problem.

Reply:
There was some discussion of this, but I think no consensus to move away from the style guidance here as it adds two different ways these need to be read with (to me at least) surprisingly similar syntax.

Jon also taught me a helpful way of reading these -- move the name inbetween the two arguments -> `std::derived_from<A, B>` means `A` is `derived_from` `B`. This is apparently a common idiom for reading and designing these kinds of APIs in Python.